[Improvement] - Publishing up/down for Menu Items

alikon opened the pull request:
Pull Request for Issue #17678 .

### Summary of Changes
add publish_up, publish_down for (site) menu items


### Testing Instructions
0. apply pr
1. fix database
2.
set publish_up, publish_down and check that (site) menu item are showed o not
accordingly to the publish_up, publish_down settings

![screenshot from 2017-08-25 08-20-07](https://user-images.githubusercontent.com/181681/29701792-416f3688-896e-11e7-841c-d48930135c9c.png)

### Expected result
you are able to set  publish_up, publish_down for (site) menu items


### Actual result
no feature


### Documentation Changes Required
?

alikon committed:
add publish_up, publish_down - mysql

add publish_up, publish_down - mysql install

alikon committed:
publish_up, publish_down blog sample

alikon committed:
publish_up, publish_down sample brochure

alikon committed:
publish_up, publish_down sample data

alikon committed:
publish_up,publish_down sample learn

alikon committed:
publish_up,publish_down sample testing

alikon committed:
publish_up,publish_down update

alikon committed:
publish_up,publish_down

alikon committed:
publish_up,publish_down

alikon committed:
publish_up, publish_down

ghost commented:
I have tested this item :red_circle: unsuccessfully on 1bac2f17beec9f513ac635c11fc401b58d78d7f6<br /><br />After apply Patch got `404 - Component not found` in Frontend on all Pages.
### System information
3.8-dev
Multilanguage Site
macOS Sierra, 10.12.6
Firefox 55 (64-bit)
#### MAMP 4.1.1
- PHP 7.0.15
- MySQLi 5.6.35<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

alikon commented:
Did you fix the database?
I'm adding 2 new fields to the menu table


On 25 Aug 2017 8:33 am, "Franz Wohlkönig" <notifications@github.com> wrote:

> I have tested this item 🔴 unsuccessfully on 1bac2f1
> <https://github.com/joomla/joomla-cms/commit/1bac2f17beec9f513ac635c11fc401b58d78d7f6>
>
> After apply Patch got 404 - Component not found in Frontend on all Pages.
> ------------------------------
> This comment was created with the J!Tracker Application
> <https://github.com/joomla/jissues> at issues.joomla.org/tracker/
> joomla-cms/17709.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/17709#issuecomment-324835550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALFse2tH_ok3b3KaSMEbCnmh-QIhdCVks5sbmq9gaJpZM4PCRNE>
> .
>

ghost commented:
after Fix Database (please update Instructions) got:
![bildschirmfoto 2017-08-25 um 08 41 51](https://user-images.githubusercontent.com/8235763/29702372-5da8b538-8971-11e7-9851-c09436a045a1.jpg)

alikon committed:
typo in table name

alikon commented:
There was a typo in table name fixed now, can you retest ;)

On 25 Aug 2017 8:42 am, "Franz Wohlkönig" <notifications@github.com> wrote:

> after Fix Database (please update Instructions) got:
> [image: bildschirmfoto 2017-08-25 um 08 41 51]
> <https://user-images.githubusercontent.com/8235763/29702372-5da8b538-8971-11e7-9851-c09436a045a1.jpg>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/17709#issuecomment-324836912>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALFsdH6HBgXEil1BtvLojtgjzhIsmIGks5sbmzwgaJpZM4PCRNE>
> .
>

brianteeman commented:
Great new feature - awesome stuff - thanks

Two small issues


1. You can set the home menu to expired or published in the future. Should we allow that as it will break the site?
2. There needs to be a status icon change in the list so you can see if it is expired etc - as you can see in the list of modules
<img width="210" alt="screenshotr07-59-25" src="https://user-images.githubusercontent.com/1296369/29702945-55fda916-896b-11e7-8a6f-0958a89c1a77.png">

ghost commented:
I have tested this item :white_check_mark: successfully on a711081a3e9a0ad7e0d8ca1757156546344d1da7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

alikon committed:
add fields to jos_menu table

add fields to jos_menu table

alikon committed:
add fileds to jos_menu.csv

add fileds to jos_menu.csv

davidsteltz commented:
I get "Error
The file marked for modification does not exist: libraries/src/Menu/SiteMenu.php' when attempting to apply the patch.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

alikon commented:
please retry maybe some synch issue

alikon committed:
postgres

alikon committed:
postgres

davidsteltz commented:
My fault, was not on latest Joomla version! Testing now...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

davidsteltz commented:
I have tested this item :white_check_mark: successfully on bb0208d2669ece3304452b9f7ec4d3debb0e1a6f<br /><br />Tested both publish and unpublish fields. Worked as expected. I agree there needs to be a status icon in the list view.

Overall great feature! Honestly, I'm shocked this hasn't been implemented sooner.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

alikon committed:
postgresql test schema

postgresql test schema

alikon committed:
sqlsrv test schema

sqlsrv test schema

alikon committed:
sqlsrv

sqlsrv

alikon committed:
sqlsrv

alikon committed:
add check for home

add check for home

alikon committed:
future check

future check

alikon committed:
publish up/down

publish up/down

alikon committed:
publish up/down

publish up/down

alikon commented:
small issues

>  You can set the home menu to expired or published in the future. Should we allow that as it will break the site?

No,

i've added a check for past/future when home, but not sure if it is enough for multilanguage

> There needs to be a status icon change in the list so you can see if it is expired etc - as you can see in the list of modules

yes, added

p.s
wasn't sure it was a good idea  :smile:

alikon committed:
up/down

alikon committed:
up\down

alikon committed:
up/down

alikon committed:
up/down

alikon committed:
up/down

alikon committed:
up/down

alikon committed:
items.

items. instead of menus.

alikon committed:
protected

protected

alikon committed:
up/down

ghost commented:
can you give a go if this PR is read for test?

alikon commented:
Is ready

On 26 Aug 2017 12:52 pm, "Franz Wohlkönig" <notifications@github.com> wrote:

> can you give a go if this PR is read for test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/17709#issuecomment-325114476>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALFsSzj-SEyZr6D95RWTGBbp69mbiKPks5sb_j4gaJpZM4PCRNE>
> .
>

ghost commented:
I have tested this item :white_check_mark: successfully on 9f57098fef0fb60c35555fd395efdc58606c233a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

infograf768 commented:
1. Any reason for tests/unit/schema/sqlsrv.sql to have windows EOL (that was also before patch)?
Also I can't apply that part of the PR with Eclipse but as it is a unit test I still can test.

2. In multilingual, the Homes can be published up and down. 
Not sure this is not a good idea.
I explain: although the flag would not display in the switcher on frontend, if the default site language is the one which Home is set with a Publish-Down in the past vs now, the page will display without a menu item and once switched to another language, we can only get back to the published down language by manually adding the url language code for that language.

I know this is already the case if we unpublish such a multilingual Home. But the multilingual status Module at least displays the error.
This is not the case when publish_down and publish_up are used.

We could modify the multilingual status Module but I guess we can simply to totally forbid any home to be saved with publish_up or down > 0.
Something like:
```
+			if ((int) $this->home)
+			{
+				// Check if publish_down or publish-up are set
+				if ((int) $this->publish_down > 0 || (int) $this->publish_up > 0)
+				{
+
+					$this->setError(JText::_('JLIB_DATABASE_ERROR_MENU_UNPUBLISH_DEFAULT_HOME'));
+
+					return false;
+				}
+			}
```
This means at least adding a new string
what we have for the the error displayed now is
`JLIB_DATABASE_ERROR_MENU_UNPUBLISH_DEFAULT_HOME="Can't unpublish default home."`
replace by something like.
`JLIB_DATABASE_ERROR_MENU_UNPUBLISH_HOME="Can't set 'Start Publishing' or 'Finish Publishing' for a Home page."`

As this would give more infos.

### This issue also shows that we do have a kind of bug in multilingual:
We should NEVER be able to display in frontend, even by manually modifying the url, any content language which Home is unpublished. In that case, the default site language could be replaced by one which has a published home, maybe by the first available one in id order.
Let's try to figure how to solve that one.

alikon committed:
multilingual can't unpublish (past/future)home

alikon commented:
home (language or not ) cannot be unpublished in the past or in the future

infograf768 commented:
@alikon 
I am now looking for a better solution for the homes. Give me some time. :)

infograf768 commented:
@alikon 
This is what I propose
1. Modify item.xml to use `showon`
```
              <field
			name="publish_up"
			type="calendar"
			label="JGLOBAL_FIELD_PUBLISH_UP_LABEL"
			description="JGLOBAL_FIELD_PUBLISH_UP_DESC"
			id="publish_up"
			class="inputbox"
			translateformat="true"
			showtime="true"
			size="22"
			filter="user_utc"
			showon="home:0"
		/>

		<field
			name="publish_down"
			type="calendar"
			label="JGLOBAL_FIELD_PUBLISH_DOWN_LABEL"
			description="JGLOBAL_FIELD_PUBLISH_DOWN_DESC"
			id="publish_down"
			class="inputbox"
			translateformat="true"
			showtime="true"
			size="22"
			filter="user_utc"
			showon="home:0"
		/>
```
Then modify menu.php to set publishdown and publishup to null
```
			if ((int) $this->home)
			{	
				$this->publish_up     = $this->_db->getNullDate();
				$this->publish_down = $this->_db->getNullDate();
			}

```

Result:
When editing an existing Home menu item, the publish down and up will not even display.
When creating a home menu item and then, before setting the item to home, a date is entered in the publish-up or down, the value will be set to Null.
In any case the value is set to Null when saving.

I think it is more simple this way and it avoids adding an error.

alikon committed:
show on when not home

show on when not home

alikon committed:
home check

home check

alikon commented:
agree,
thanks @infograf768

infograf768 commented:
I have tested this item :white_check_mark: successfully on 9f57098fef0fb60c35555fd395efdc58606c233a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

ghost commented:
RTC after two successful tests.

infograf768 commented:
@mbabker 
Please decide the milestone.

mbabker commented on administrator/components/com_menus/models/items.php:
```diff
@@ -234,12 +234,14 @@ protected function getListQuery()
 						null, null, null, null, null, null, null, null, null,
-						null, 'a.published', null, null, null, null,
-						null, null, null, null, null, null, null, null, null
+						null, null, null, null, null, null,
```
Do we even need this second array now that there isn't anything aliased in the column list?

alikon replied:
no,
only forgot to remove

mbabker commented on administrator/components/com_menus/tables/menu.php:
```diff
@@ -43,4 +43,50 @@ public function delete($pk = null, $children = false)
+
+			}
+		
+			return $return;
```
This return should be outside the if statement.

alikon replied:
ops  :innocent:

mbabker commented on libraries/src/Menu/SiteMenu.php:
```diff
@@ -72,9 +72,10 @@ public function __construct($options = array())
 		// For PHP 5.3 compat we can't use $this in the lambda function below
-		$db = $this->db;
+		$db          = $this->db;
+		$currentDate = \JFactory::getDate()->toSql();
```
Is there a reason this can't be set inside the lambda function?  It's not used anywhere else.

alikon replied:
moved inside
thanks

alikon committed:
no more alias

no more alias

alikon committed:
fix wrong place for return

fix wrong place for return

alikon committed:
move inside lambda()

move inside lambda()

infograf768 commented:
Setting back to pending as we need new tests.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

infograf768 commented:
Some issues here
Time Zone set to UTC in global config.
User set to use default.

Start publishing = today
Finishing publishing = tomorrow.

Dates displayed in the fields is French time (CEST) = UTC+2

![screen shot 2017-08-30 at 09 35 32](https://user-images.githubusercontent.com/869724/29860595-da29c108-8d66-11e7-9244-a2aa434e0023.png)


![screen shot 2017-08-30 at 09 27 38](https://user-images.githubusercontent.com/869724/29860230-a524ed08-8d65-11e7-9ba2-f7497c64ce04.png)

![screen shot 2017-08-30 at 09 19 55](https://user-images.githubusercontent.com/869724/29860252-c61d1ef4-8d65-11e7-8146-69e8c69923a0.png)

The result is that the menu item does not display although it should.

![screen shot 2017-08-30 at 09 22 06](https://user-images.githubusercontent.com/869724/29860244-bddfd8a8-8d65-11e7-94b8-40fee432a473.png)

After clearing the publish up and down, I do get correctly

![screen shot 2017-08-30 at 09 40 51](https://user-images.githubusercontent.com/869724/29860781-72dd3268-8d67-11e7-8746-3c48e6a3c66f.png)

infograf768 commented:
hmm, it looks like the issue is unrelated to this PR.
@dgt41 @mbabker 
Please look at this. To test, use UTC in global config, clear the start publishing date of an article, save, then click on today.
The article is set Published but  pending

![screen shot 2017-08-30 at 09 54 37](https://user-images.githubusercontent.com/869724/29861375-4e33e356-8d69-11e7-9806-c50cde22fc37.png)

brianteeman commented:
@infograf768 cant find it now but i am pretty sure we have an open issue about this exact thing

infograf768 commented:
As I can't find it either, I create a new issue.

infograf768 commented:
See https://github.com/joomla/joomla-cms/issues/17770

infograf768 commented:
Apart from that issue, the PR looks good.

infograf768 commented:
I have tested this item :white_check_mark: successfully on 9f57098fef0fb60c35555fd395efdc58606c233a<br /><br />As the issue is a general issue in core (2.5.28 does not have this problem), this PR is OK for me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

ghost commented:
I have tested this item :white_check_mark: successfully on 9f57098fef0fb60c35555fd395efdc58606c233a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

ghost commented:
RTC after two successful tests.

Quy commented on administrator/components/com_menus/tables/menu.php:
```diff
@@ -43,4 +43,50 @@ public function delete($pk = null, $children = false)
+				// Set the publish down/up always for home.
+				$this->publish_up   = $this->_db->getNullDate();
+				$this->publish_down = $this->_db->getNullDate();
+
```
Remove blank line

alikon replied:
removed

Quy commented on libraries/src/Menu/SiteMenu.php:
```diff
@@ -72,17 +72,20 @@ public function __construct($options = array())
 	{
 		// For PHP 5.3 compat we can't use $this in the lambda function below
-		$db = $this->db;
+		$db          = $this->db;
```
Why extra spaces?

alikon replied:
removed

alikon committed:
cs

cs

alikon committed:
cs

cs

alikon commented:
hope no need to retest, only cosmetic change

ghost commented:
@alikon it's an easy Test :-)

Quy commented:
URL menu item type does not have this option. Intentional?

infograf768 commented:
> URL menu item type does not have this option. Intentional?

Looks like it concerns all System Links.

infograf768 commented:
Found out why.
For these menutypes, we do not display the Home field.
As displaying the publish-up and publish-down fields depends on the setting of the Home (Default Page) to No, they will not display.
Code to not display the Home field for system links is in `/administrator/components/com_menus/views/item/tmpl/edit.php` line 150-153.
```
				if ($this->item->type != 'component')
				{
					$this->fields = array_diff($this->fields, array('home'));
				}
```

alikon commented:
Good catch @Quy, it wasn't intentional
@infograf768 can you do the fix ?

On 31 Aug 2017 9:58 am, "infograf768" <notifications@github.com> wrote:

> Found out why.
> For these menutypes, we do not display the Home field.
> As displaying the publish-up and publish-down fields depends on the
> setting of the Home (Default Page) to Yes, they will not display.
> Code to not display the Home field for system links is in
> /administrator/components/com_menus/views/item/tmpl/edit.php line 150-153.
>
> 				if ($this->item->type != 'component')
> 				{
> 					$this->fields = array_diff($this->fields, array('home'));
> 				}
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/17709#issuecomment-326220372>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALFsbGXIiBZh3p3VCtLpLcb_1e5LWeaks5sdme8gaJpZM4PCRNE>
> .
>

infograf768 commented:
The fix should be to add in` /administrator/components/com_menus/views/item/view.html.php` line 79.

Add

```
		// System Links have no home. Force display of publish_up and publish_down.
		if ($this->item->type == 'alias' || $this->item->type == 'url'
			|| $this->item->type == 'separator' || $this->item->type == 'heading')
		{
			$this->form->setFieldAttribute('publish_up', 'showon', '');
			$this->form->setFieldAttribute('publish_down', 'showon', '');
		}
```

infograf768 commented:
@alikon 
Can you modify yourself?

infograf768 commented:
Can be added in fact in edit.php.

infograf768 committed:
Force publish-up and down on system lunks

infograf768 commented:
@alikon 
Committed directly.

@Quy 
Can you test?

alikon commented:
Thanks @infograf768

On 31 Aug 2017 10:25 am, "infograf768" <notifications@github.com> wrote:

@alikon <https://github.com/alikon>
Committed directly.

@Quy <https://github.com/quy>
Can you test?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/joomla/joomla-cms/pull/17709#issuecomment-326226672>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AALFsUWOZRGwbcdLwnPzw0esyM7eg2Ttks5sdm4OgaJpZM4PCRNE>
.

infograf768 commented:
@mbabker 
Can we get this improvement in 3.8.0 ?

jsubri commented:
I have tested this item :white_check_mark: successfully on 9f57098fef0fb60c35555fd395efdc58606c233a<br /><br />I've tested unpublished menu such as "404 not Found" with date in past/future. Working as before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17709">issues.joomla.org/tracker/joomla-cms/17709</a>.</sub>

TLWebdesign commented:
I see there hasn't been any activity here. Was wondering what the status is and when we can expect this feature?

ghost commented:
@TLWebdesign after the needed 2 successfully Tests its a Decision of Release Lead (3.8: @mbabker)

brianteeman commented:
This needs to be tested to see what happens if a menu type of single article (for example) is unpublished and the results of a search tries to load that article

alikon commented:
good point
i've made some test with `com_search` only, 
and the article are searched & loaded indipendently if the article is an item of a menu and that menu is not showed cause of publish/unpubish datetime, should be the expected behaviour imo
but i agree more tests are always better

p.s
com_finder should behave the same

brianteeman commented:
yeah thats what i thought. i think users will expect it not to be found if the menu item is unpublished. otherwise they have to unpublish the article and the menu item but they will think that the menu is enough - i know i would have done

alikon commented:
as it is a new feature  should be enough to details its (logical imho) beahviuor in the documentation ?

csthomas commented:
I see some things that should be described:
- if we use cache then menu will be unpublished (in the worst case after publish_down + cache timeout)
- for the date after publication (publish_down), the menu item will not be visible in the module menu but links added manual in other places (article) will be changed (after cache timeout), ex: 
    * from `/category-menu/article-menu-item-with-publish-down`
    * to `/category-menu/123-article`
    * or to `/component/content/1-category/123-article`

zero-24 commented:
@alikon can you please fix the merge conflicts?

uglyeoin commented:
Is this still in the making?

pe7er committed:
Fixed small typos in months of Joomla core extension (#23552)

brianteeman committed:
CodeStyle - use caps (#23558)

All language constants should be in uppercase - some where lowercase in error

brianteeman committed:
CSS Class for tag link (#23561)

Update string to be more accurate and consistent with similar strings in template styles

wojsmol committed:
Update item.xml (#23556)

brianteeman committed:
Smart Search (finder) module (#23560)

### Expected
The module  form should be wrapped in a div that a class suffix can be applied to

### Actual
The module div is wrapped in the form tag

### Changes
This simple PR switches the form and the div - making it the same structure as other modules - see mod_search

alikon committed:
[com_actionlogs] - log checkin actions (#23526)

* [com_ajax] - need to be protected

postgresql update

* log checkin

* udpate

* old one

* update fixes

Hackwar committed:
Deprecating loadFieldType & loadRuleType (#23478)

alikon committed:
featured view fix filter on category (#23454)

csthomas committed:
Fix sql queries in the update file for SQL Server (#23210)

infograf768 committed:
Debug language correction (#23590)

alikon committed:
Merge branch '3.10-dev' into patch-95

alikon committed:
Merge branch '3.10-dev' into patch-95

alikon commented:
> Is this still in the making?
moved on 3.10 branch let's see :smile:

uglyeoin commented:
I think this is such a useful pull request, if you need someone to test it please tag me and I will test it for you @alikon .  Let me know whenever you are ready.

TLWebdesign commented:
This would be very usefull indeed. I’m no help besides testing but jeep up the good work.

franz-wohlkoenig commented:
@HLeithner This Pull Request have Status "Ready for Commit" and conflicting Files, so Status removed.

franz-wohlkoenig commented:
But now the 3.10-Milestone is gone.

jeckodevelopment commented:
@alikon Nicola can you please fix conflicts?


Conflicting files
```
installation/sql/mysql/sample_brochure.sql
installation/sql/mysql/sample_learn.sql
installation/sql/mysql/sample_testing.sql
tests/unit/stubs/database/jos_menu.csv
```

franz-wohlkoenig commented:
@jeckodevelopment can you reassign 3.10-Milestone?

jeckodevelopment commented:
generally milestone is assigned once in RTC status

alikon commented:
i'll be happy to fix conflicts, just want to know, where new feature like this should be supposed to land ?

- 3.9.x 
- 3.10.x 
- 4.x 

p.s
it's surfing the joomla sea from 3.8.0

zero-24 commented:
@HLeithner as RL should be able to help with that decision 👍

HLeithner commented:
@alikon please rebase on 4.0, I watched @wilsonge merging 3.9 commits in to 4.0 branch....

alikon commented:
done  #24416
